crosschain-resolver/contracts/L1Resolver.sol

```diff
@@ -4,20 +4,62 @@ pragma solidity ^0.8.17;
 import {EVMFetcher} from '@ensdomains/evm-verifier/contracts/EVMFetcher.sol';
 import {EVMFetchTarget} from '@ensdomains/evm-verifier/contracts/EVMFetchTarget.sol';
 import {IEVMVerifier} from '@ensdomains/evm-verifier/contracts/IEVMVerifier.sol';
+import "@ensdomains/ens-contracts/contracts/registry/ENS.sol";
+import {INameWrapper} from "@ensdomains/ens-contracts/contracts/wrapper/INameWrapper.sol";
 
 contract L1Resolver is EVMFetchTarget {
     using EVMFetcher for EVMFetcher.EVMFetchRequest;
     IEVMVerifier immutable verifier;
-    address immutable target;
+    ENS immutable ens;
+    INameWrapper immutable nameWrapper;
+    mapping(bytes32 => address) public targets;
     uint256 constant COIN_TYPE_ETH = 60;
-    uint256 constant RECORD_VERSIONS_SLOT = 1;
-    uint256 constant VERSINABLE_ADDRESSES_SLOT = 3;
-    uint256 constant VERSINABLE_HASHES_SLOT = 4;
-    uint256 constant VERSINABLE_TEXTS_SLOT = 11;
-
-    constructor(IEVMVerifier _verifier, address _target) {
-        verifier = _verifier;
-        target = _target;
+    uint256 constant RECORD_VERSIONS_SLOT = 0;
+    uint256 constant VERSINABLE_ADDRESSES_SLOT = 2;
+    uint256 constant VERSINABLE_HASHES_SLOT = 3;
+    uint256 constant VERSINABLE_TEXTS_SLOT = 10;
+
+    event TargetSet(bytes32 indexed node, address target);
+
+    function isAuthorised(bytes32 node) internal view returns (bool) {
+        // TODO: Add support for
+        // trustedETHController
+        // trustedReverseRegistrar
+        // isApprovedForAll
+        // isApprovedFor
+        address owner = ens.owner(node);
+        if (owner == address(nameWrapper)) {
+          owner = nameWrapper.ownerOf(uint256(node));
+        }
+        return owner == msg.sender;
+    }
+
+    modifier authorised(bytes32 node) {
+        require(isAuthorised(node));
+        _;
+    }
+
+    constructor(
+      IEVMVerifier _verifier,
+      ENS _ens
+      ,INameWrapper wrapperAddress
+    ){
+      require(address(wrapperAddress) != address(0), "Name Wrapper address must be set");
+      require(address(_verifier) != address(0), "Verifier address must be set");
+      require(address(_ens)  != address(0), "Registry address must be set");
+      verifier = _verifier;
+      ens = _ens;
+      nameWrapper = wrapperAddress;
+    }
+
+    /**
+     * Set target address to verify aagainst
+     * @param node The ENS node to query.
+     * @param _target The L2 resolver address to verify against.
+     */
+    function setTarget(bytes32 node, address _target) public authorised(node){
```

Comment:
Doesn't having to set a target per address remove a lot of the point of moving your name to an L2 like this? It makes wildcard subdomains impossible via this resolver. If we're going to make one deployment work for multiple targets, it should at least use `resolve` and search parent names for a target, like the `DelegatableResolver` does.